### DISCUSSION TITLE
macOS - enabling FIPS

### DISCUSSION THREAD

**Discussion by szigetics** [openssl/openssl#25836] [Q&A] [answered]
I am working on enabling the FIPS module in my macOS app.

**Briefly my question is the following :** 
I need to be able to sign the fips.dylib file before the fipsmodule.cnf gets generated.
How could I add the relevant `codesign --timestamp --options runtime -f -s "<my identity> fips.dylib"` call before the fipsmodule.cnf gets generated?

**More details :** 
I am using OpenSSL version 3.

Here is the reason why I need to sign the file : 
Using my current fips.dylib build, which is only ad-hoc signed, everything works well if I disable the SIP (system integrity protection).
When it's enabled, loading the fips module fails with the following error : 
> Error : 400FEFFB01000000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:
crypto/dso/dso_dlfcn.c:118:
> 400FEFFB01000000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:147:
> 400FEFFB01000000:error:07880025:common libcrypto routines:provider_init:reason(37):crypto/provider_core.c:949:name=fips_armv8

Apple's WatchDog is rejecting the dylib, because it is not signed : 
> spctl --assess --verbose=4 fips_armv8.dylib
> fips_armv8.dylib: rejected

(
Side notes : 
1.: I just simply renamed the fips.dylib that was generated by the build to fips_armv8.dylib.
2.: I used the usual build steps for building the openssl libs and fips.dylib : 
```
./config --strict-warnings --release enable-fips --prefix=/tmp/mybuild darwin64-arm64-cc
make clean
make -j8
make install_sw
make install_fips
```
)

**Comment by tom-cosgrove-arm** [comment]
> I am using OpenSSL version 3.

version 3 point what? 3.0.1, 3.1.5, 3.2.2, 3.3.1, 3.4.0, or something else?

> I just simply renamed the fips.dylib that was generated by the build to fips_armv8.dylib.

Could this be part of the problem? There are internal IDs in `dylib`s which don't get changed when the file is renamed

**Comment by szigetics** [comment]
@tom-cosgrove-arm  : 
> > I am using OpenSSL version 3.
> 
> version 3 point what? 3.0.1, 3.1.5, 3.2.2, 3.3.1, 3.4.0, or something else?
> 
> > I just simply renamed the fips.dylib that was generated by the build to fips_armv8.dylib.
> 
> Could this be part of the problem? There are internal IDs in `dylib`s which don't get changed when the file is renamed

I am using 3.3.2 at the moment, but I can switch to any newer version if that would help fixing the issue.

No, that cannot be the problem : as I mentioned, without SIP (macOS system integrity protection) enabled, everything is working well. This (and the error message I quoted) proves that the issue is only related to SIP (or GateKeeper) blocking the loading of dylib which is not signed.

This is the reason why I would need to add a `codesign` right before the fipsmodule.cnf file is being generated.
So I am digging the [Configure](https://github.com/openssl/openssl/blob/openssl-3.3/Configure) and the [unix-Makefile.tmpl](https://github.com/openssl/openssl/blob/openssl-3.3/Configurations/unix-Makefile.tmpl) right now to check where/how could I add the `codesign` step...

**Comment by mattcaswell** [comment]
> before the fipsmodule.cnf gets generated?

Why not just regenerate fipsmodule.cnf after the signing step via the fipsinstall app?

https://docs.openssl.org/master/man1/openssl-fipsinstall/#name

**Reply by szigetics** [reply]
@mattcaswell  : 
Calling `openssl fipsinstall` works well when I build from x86_64 to x86 _64, but it doesn't work when I am doing cross compilation from x86_64 to arm64.

(Reason : the `openssl` is built for arm64, so it cannot be launched on an x86_64 machine.)

**Reply by szigetics** [reply, marked as answer]
I found a way to achieve what I needed using the normal build process : 
https://github.com/openssl/openssl/blob/openssl-3.3/util/mk-fipsmodule-cnf.pl
needed to be patched like this : 
```
diff --git a/util/mk-fipsmodule-cnf.pl b/util/mk-fipsmodule-cnf.pl
index 6a86e06b8b..419366f741 100644
--- a/util/mk-fipsmodule-cnf.pl
+++ b/util/mk-fipsmodule-cnf.pl
@@ -22,6 +22,13 @@ GetOptions("key=s"              => \$mac_key,
 
 my $mac_keylen = length($mac_key);
 
+# Sign the module with codesign on macOS
+if ($^O eq 'darwin') {  # Check if running on macOS
+    my $identity = "my code signing identity"; # Replace with your code signing identity
+    system("codesign", "--timestamp", "--options", "runtime", "--force", "--sign", $identity, $module_name) == 0
+        or die "Failed to codesign $module_name: $!";
+}
+
 use Digest::SHA qw(hmac_sha256_hex);
 my $module_size = [ stat($module_name) ]->[7];
 

```

After this modification I can just simply call the build commands as normal : 
```
./config --strict-warnings --debug enable-fips --prefix=/Users/cszigeti/Downloads/openssl-3.3.2/nxt_build darwin64-arm64-cc
make clean
make -j8
make install_sw
make install_fips
```
